File: client/src/api/stocks.ts
import { fetchJson, getApiBaseUrl, toQueryString } from "./utils";
import { QuoteInterval, ApiResult, SymbolsApiResponse, ServerStockInfo } from "./types";
import { StockPerformance } from "./types";

const API_BASE_URL = getApiBaseUrl();
const STOCKS_BASE = `${API_BASE_URL}/api/v1/stocks`;

export const stocksApi = {
  urls: {
    quote: (
      symbol: string,
      params?: { startDate?: string; endDate?: string; interval?: QuoteInterval }
    ) =>
      `${STOCKS_BASE}/quote?symbol=${encodeURIComponent(symbol)}${toQueryString({
        startDate: params?.startDate,
        endDate: params?.endDate,
        interval: params?.interval,
      })}`,
    intraday: (symbol: string) =>
      `${STOCKS_BASE}/intraday?symbol=${encodeURIComponent(symbol)}`,
    priceDepth: (symbol: string) =>
      `${STOCKS_BASE}/price-depth?symbol=${encodeURIComponent(symbol)}`,
    company: (symbol: string) => `${STOCKS_BASE}/company?symbol=${encodeURIComponent(symbol)}`,
    shareholders: (symbol: string) =>
      `${STOCKS_BASE}/shareholders?symbol=${encodeURIComponent(symbol)}`,
    priceBoard: (symbols: string[]) =>
      `${STOCKS_BASE}/price-board${toQueryString({
        symbols: symbols.map((s) => s.trim()).filter(Boolean).join(","),
      })}`,
    stockInfo: (symbols: string[]) =>
      `${STOCKS_BASE}/stock-info${toQueryString({
        symbols: symbols.map((s) => s.trim()).filter(Boolean).join(","),
      })}`,
    stockPerformance: (symbol: string) =>
      `${STOCKS_BASE}/stock-performance${toQueryString({ symbol })}`,
    symbols: () => `${STOCKS_BASE}/all-symbols`,
    indices: () => `${STOCKS_BASE}/indices`,
  },

  /**
   * React-query friendly query keys.
   * Use as: useQuery({ queryKey: stocksApi.queryKeys.quote("HPG"), ... })
   */
  queryKeys: {
    quote: (
      symbol: string,
      params?: { startDate?: string; endDate?: string; interval?: QuoteInterval }
    ) => [stocksApi.urls.quote(symbol, params)] as const,
    intraday: (symbol: string) => [stocksApi.urls.intraday(symbol)] as const,
    priceDepth: (symbol: string) => [stocksApi.urls.priceDepth(symbol)] as const,
    company: (symbol: string) => [stocksApi.urls.company(symbol)] as const,
    shareholders: (symbol: string) => [stocksApi.urls.shareholders(symbol)] as const,
    priceBoard: (symbols: string[]) => [stocksApi.urls.priceBoard(symbols)] as const,
    symbols: () => [stocksApi.urls.symbols()] as const,
    indices: () => [stocksApi.urls.indices()] as const,
    stockPerformance: (symbol: string) => [stocksApi.urls.stockPerformance(symbol)] as const,
  },

  getQuote: <T = unknown>(
    symbol: string,
    params?: { startDate?: string; endDate?: string; interval?: QuoteInterval },
    options?: { signal?: AbortSignal }
  ) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.quote(symbol, params), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getIntraday: <T = unknown>(symbol: string, options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.intraday(symbol), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getPriceDepth: <T = unknown>(symbol: string, options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.priceDepth(symbol), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getCompanyInfo: <T = unknown>(symbol: string, options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.company(symbol), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getShareholders: <T = unknown>(symbol: string, options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.shareholders(symbol), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getPriceBoard: <T = unknown>(symbols: string[], options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.priceBoard(symbols), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getStockInfo: <T = ServerStockInfo>(symbols: string[], options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.stockInfo(symbols), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getAllSymbols: <T = SymbolsApiResponse>(options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.symbols(), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getIndices: <T = unknown>(options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.indices(), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),

  getStockPerformance: <T = StockPerformance>(symbol: string, options?: { signal?: AbortSignal }) =>
    fetchJson<ApiResult<T>>(stocksApi.urls.stockPerformance(symbol), {
      signal: options?.signal,
      logPrefix: "[stocksApi]",
    }),
};

File: server/app/api/v1/routes/stocks.py
from __future__ import annotations

from typing import Optional, List

from fastapi import APIRouter, Query, HTTPException, Depends
from sqlalchemy.orm import Session

from app.services import stocks_service
from app.db.session import get_db
from app.db.models import Watchlist

router = APIRouter(prefix="/stocks", tags=["stocks"])


@router.get("/quote")
def quote(
    symbol: str = Query(..., min_length=1),
    startDate: Optional[str] = None,
    endDate: Optional[str] = None,
    interval: str = "1D",
):
    result = stocks_service.get_stock_quote(symbol, startDate, endDate, interval)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/intraday")
def intraday(symbol: str = Query(..., min_length=1)):
    result = stocks_service.get_stock_intraday(symbol)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/price-depth")
def price_depth(symbol: str = Query(..., min_length=1)):
    result = stocks_service.get_stock_price_depth(symbol)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/company")
def company(symbol: str = Query(..., min_length=1)):
    result = stocks_service.get_company_info(symbol)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/shareholders")
def shareholders(symbol: str = Query(..., min_length=1)):
    result = stocks_service.get_company_shareholders(symbol)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/price-board")
def price_board(symbols: str = Query(..., description="Comma-separated symbols, e.g. VNM,HPG")):
    parsed: List[str] = [s.strip() for s in symbols.split(",")]
    # print("Parsed symbols:", parsed)
    result = stocks_service.get_price_board(parsed)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result

@router.get("/stock-info")
def stock_info(symbols: str = Query(..., description="Comma-separated symbols, e.g. VNM,HPG")):
    parsed: List[str] = [s.strip() for s in symbols.split(",")]
    print("Parsed symbols:", parsed)
    result = stocks_service.get_stock_info(parsed)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result

@router.get("/stock-performance")
def stock_performance(symbol: str = Query(..., min_length=1)):
    result = stocks_service.get_stock_performance(symbol)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result

@router.get("/dividend-history")
def dividend_history(symbol: str = Query(..., min_length=1)):
    result = stocks_service.get_stock_dividends(symbol)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result

@router.get("/all-symbols")
def all_symbols():
    result = stocks_service.get_all_symbols()
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/watchlist")
def get_watchlist(db: Session = Depends(get_db)):
    """
    Get all stocks in the watchlist with their stock info.
    These stocks are monitored by the background analysis worker.
    """
    try:
        watchlist_items = db.query(Watchlist).filter_by(user_id=None).all()
        
        # Get symbols
        symbols = [item.symbol for item in watchlist_items]
        
        # Fetch stock info for all symbols in one call
        stock_info_result = {}
        if symbols:
            stock_info_response = stocks_service.get_stock_info(symbols)
            if "error" not in stock_info_response and "data" in stock_info_response:
                # stock_info_response["data"] is a dict with symbol keys
                # e.g., {"HPG": {...}, "VNM": {...}}
                stock_info_result = stock_info_response["data"]
        
        return {
            "status": "success",
            "data": [
                {
                    "id": item.id,
                    "symbol": item.symbol,
                    "added_at": item.added_at.isoformat(),
                    "stock_info": stock_info_result.get(item.symbol)
                }
                for item in watchlist_items
            ],
            "total": len(watchlist_items)
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/watchlist")
def add_to_watchlist(symbol: str = Query(..., min_length=1), db: Session = Depends(get_db)):
    """
    Add a stock to the watchlist for background monitoring.
    """
    try:
        # Check if already exists
        existing = db.query(Watchlist).filter_by(symbol=symbol, user_id=None).first()
        if existing:
            raise HTTPException(status_code=400, detail=f"{symbol} is already in watchlist")
        
        # Add to watchlist
        watchlist_item = Watchlist(symbol=symbol, user_id=None)
        db.add(watchlist_item)
        db.commit()
        db.refresh(watchlist_item)
        
        # Fetch stock info for the added symbol
        stock_info_response = stocks_service.get_stock_info([symbol])
        stock_info = None
        if "error" not in stock_info_response and "data" in stock_info_response:
            # stock_info_response["data"] is a dict with symbol keys
            # e.g., {"HPG": {...}}
            stock_info = stock_info_response["data"].get(symbol)
        
        return {
            "status": "success",
            "message": f"{symbol} added to watchlist",
            "data": {
                "id": watchlist_item.id,
                "symbol": watchlist_item.symbol,
                "added_at": watchlist_item.added_at.isoformat(),
                "stock_info": stock_info
            }
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/watchlist/{symbol}")
def remove_from_watchlist(symbol: str, db: Session = Depends(get_db)):
    """
    Remove a stock from the watchlist.
    """
    try:
        watchlist_item = db.query(Watchlist).filter_by(symbol=symbol, user_id=None).first()
        if not watchlist_item:
            raise HTTPException(status_code=404, detail=f"{symbol} not found in watchlist")
        
        db.delete(watchlist_item)
        db.commit()
        
        return {
            "status": "success",
            "message": f"{symbol} removed from watchlist"
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/indices")
def indices():
    # indices returns empty array if none, not an error
    return stocks_service.get_indices()